pkg/bff/mock/admin.go

```diff
@@ -0,0 +1,162 @@
+// UseHandler methods.
+type Admin struct {
+	srv       *httptest.Server
+	router    *gin.Engine
+	client    apiv2.DirectoryAdministrationClient
+	tokens    *tokens.TokenManager
+	authorize gin.HandlerFunc
+	csrf      gin.HandlerFunc
+	handlers  map[string]gin.HandlerFunc
+	Calls     map[string]int
+}
+
+// UseFixture allows tests to use a JSON fixture from disk as the response.
+func (a *Admin) UseFixture(endpoint, path string) (err error) {
+	// Read the fixture data from disk
+	var data []byte
+	if data, err = ioutil.ReadFile(path); err != nil {
+		return fmt.Errorf("could not read fixture data: %s", err)
+	}
+
+	switch endpoint {
+	case StatusEP:
+		out := &apiv2.StatusReply{}
+		if err = json.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal fixture data: %s", err)
+		}
+		if err = a.handle(endpoint, func(c *gin.Context) {
+			c.JSON(http.StatusOK, out)
+		}); err != nil {
+			return fmt.Errorf("could not handle fixture data: %s", err)
+		}
+	case ListCertificatesEP:
+		out := &apiv2.ListCertificatesReply{}
+		if err = json.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal fixture data: %s", err)
+		}
+		if err = a.handle(endpoint, func(c *gin.Context) {
+			c.JSON(http.StatusOK, out)
+		}); err != nil {
+			return fmt.Errorf("could not set handler: %s", err)
+		}
+	default:
+		return fmt.Errorf("unsupported endpoint: %s", endpoint)
+	}
+
+	return nil
+}
+
+// UseError allows tests to specify an HTTP error for the response.
+func (a *Admin) UseError(endpoint string, code int, msg string) (err error) {
+	return a.handle(endpoint, func(c *gin.Context) {
+		c.JSON(code, msg)
+	})
+}
+
+// UseHandler allows tests to specify a custom handler for the endpoint.
+func (a *Admin) UseHandler(endpoint string, handler gin.HandlerFunc) (err error) {
+	return a.handle(endpoint, handler)
+}
+
+// handle is a helper function that adds a handler to the handlers map and
+// returns that handler function when the endpoint is called.
+func (a *Admin) handle(endpoint string, handler gin.HandlerFunc) error {
+	if _, ok := a.handlers[endpoint]; !ok {
+		return fmt.Errorf("unhandled endpoint %s", endpoint)
+	}
+	a.handlers[endpoint] = handler
+	return nil
+}
+
+// initHandler initializes a handler for the given endpoint on the gin router with the
+// provided method, path, and middleware. After calling this function the configured
+// handler for the endpoint will return a 200 and an empty response. This should only
+// be called once for each endpoint to avoid a gin panic. UseFixture, UseError, or
+// UseHandler should be used to configure different endpoint responses from the tests.
+func (a *Admin) initHandler(endpoint, method, path string, middleware ...gin.HandlerFunc) {
+	var handlers []gin.HandlerFunc
+
+	a.handlers[endpoint] = func(c *gin.Context) {
+		c.JSON(http.StatusOK, gin.H{})
+	}
+
+	if len(middleware) > 0 {
+		handlers = append(handlers, middleware...)
+	}
```

Comment:
The len check is a bit unnecessary since you're just making a copy of the slice; e.g. you could do:

```go
handlers := make([]gin.HandlerFunc, len(middleware), len(middleware)+1)
copy(handlers, middleware)

… 
handlers = append(handlers, handlerFunc) 
```

I do prefer this to the `len` check and `var` declaration because it is more explicit about what the code is doing, you're making a copy -- and that's important, because if you did `middleware = append(middleware, handlerFunc); a.router.Handle(method, path, middleware…)` then on the next call to `initHandler` you might have an extra "middleware" (though that won't happen the way this code is written).